### PR TITLE
[Merged by Bors] - refactor(Mathport): move mathport prelude to mathlib4

### PR DESCRIPTION
**Pull request by kim-em**
Moving `Mathport/Prelude/*` out of mathport into mathlib4. I've added a README that will hopefully help bystanders:
```
Mathport prelude
===

This directory contains instructions for `mathport`,
helping it to translate `mathlib` and align declarations and tactics with `mathlib4`.
(These files were formerly part of `mathport`, in the directory `Mathport/Prelude/`.)

`SpecialNames.lean`
: Contains `#align X Y` statements, where `X` is an identifier from mathlib3
  which should be aligned with the identifier `Y` from mathlib4.
  Sometimes we need `#align` statements just to handle exceptions to casing rules,
  but there are also many exceptional cases.

`Syntax.lean`
: Contains unimplemented stubs of tactics which need to be migrated from Lean3 to Lean4.
  When porting tactics, you can move the relevant stubs to a new file and
  use them as a starting point.
  Please make sure this file stays in sync with new tactic implementations
  (and in particular that the syntax is not defined twice).
  Please preserve the syntax of existing mathlib tactics,
  so that there are no unnecessary parse errors in the source files generated by `synport`.
  It is fine to fail with an error message for unimplemented features for now.
```

### CODE REVIEW

**Review by digama0 — COMMENTED** [review]
Most of the commented out tactics are not actually a match, so the mathlib4 tactic needs to be adjusted to fit, or else we should delay replacing the tactic stub and importing the new tactic.

The other thing that should potentially cause build errors in mathport is if the mathlib4 tactic has a different name than these tactics. Most of the mathlib tactics are not named, but I think we should keep a standard naming convention (and placing them in the `Lean.Tactic.foo` namespace, as is done here) to make it easier to detect collisions between lean4 and mathlib4 tactics, or mathlib4 tactics and mathport tactic stubs.

File: Mathlib/Tactic/ShowTerm.lean [hunk 1]
```diff
@@ -13,13 +13,15 @@ open Tactic.TryThis
 
 namespace Lean.Elab.Tactic
 
+syntax (name := showTerm) "showTerm " tacticSeq : tactic
+
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Tactic/ShowTerm.lean]: Okay --- I have remove the separate `syntax` statement, and changed the `elab` line to:
```
elab (name := showTerm) tk:"showTerm " t:tacticSeq : tactic => 
```

File: Mathlib/Mathport/Syntax.lean [hunk 2]
```diff
@@ -0,0 +1,737 @@
...
+  (" [" simpArg,* "]")? (" with " (colGt ident)+)? (" using " (colGt ident)+)? : tactic
+
+syntax (name := tauto) "tauto" (" (" &"config" " := " term ")")? : tactic
+syntax (name := tauto!) "tauto!" (" (" &"config" " := " term ")")? : tactic
+
+syntax (name := truncCases) "truncCases " term (" with " (colGt binderIdent)+)? : tactic
+
+syntax (name := normNum1) "normNum1" (ppSpace location)? : tactic
+-- Implemented in Mathlib.Tactic.NormNum
+-- syntax (name := normNum) "normNum" (" [" simpArg,* "]")? (ppSpace location)? : tactic
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Mathport/Syntax.lean:464]: I've tried declaring the full syntax, and then using an `elab_rule` to match just `normNum` by itself, and as far as I can see these declarations are working, but at `by normNum` I just get `tactic 'Lean.Tactic.normNum' has not been implemented`.

@digama0, do you know what I'm doing wrong with `elab_rules` here?

File: Mathlib/Mathport/Syntax.lean [hunk 3]
```diff
@@ -0,0 +1,737 @@
...
+syntax (name := listLinters) "#list_linters" : command
+
+syntax (name := copyDocString) "copy_doc_string " ident " → " ident* : command
+syntax (name := libraryNote) docComment "library_note " str : command
+syntax (name := addTacticDoc) (docComment)? "add_tactic_doc " term : command
+syntax (name := addDeclDoc) docComment "add_decl_doc " ident : command
+
+syntax (name := setupTacticParser) "setup_tactic_parser" : command
+-- See Mathlib.Tactic.OpenPrivate, which provides an alternative:
+-- syntax (name := importPrivate) "import_private " ident (" from " ident)? : command
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Mathport/Syntax.lean]: Okay, I've left this one as is.

In https://github.com/leanprover-community/mathlib/pull/10084 I removed the only use of `import_private` from mathlib3, to simplify life.

File: Mathlib/Mathport/Syntax.lean [hunk 4]
```diff
@@ -0,0 +1,737 @@
...
+syntax (name := squeezeDSimp?!) "squeezeDSimp?!" (" (" &"config" " := " term ")")? (&" only")?
+  (" [" simpArg,* "]")? (" with " (colGt ident)+)? (ppSpace location)? : tactic
+
+syntax (name := suggest) "suggest" (" (" &"config" " := " term ")")? (ppSpace num)?
+  (" [" simpArg,* "]")? (" with " (colGt ident)+)? (" using " (colGt ident)+)? : tactic
+-- Implemented in Mathlib.Tactic.LibrarySearch
+-- syntax (name := librarySearch) "librarySearch" (" (" &"config" " := " term ")")?
+--   (" [" simpArg,* "]")? (" with " (colGt ident)+)? (" using " (colGt ident)+)? : tactic
+syntax (name := librarySearch!) "librarySearch!" (" (" &"config" " := " term ")")?
+  (" [" simpArg,* "]")? (" with " (colGt ident)+)? (" using " (colGt ident)+)? : tactic
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Mathport/Syntax.lean]: Unfortunately with `elab_rules` I don't know how to grab the `Syntax` term that `tk` was previously binding. I've asked on [zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Binding.20the.20syntax.20being.20matched.20in.20.60elab_rules.60.3F/near/259825788).

File: Mathlib/Mathport/Syntax.lean [hunk 5]
```diff
@@ -0,0 +1,737 @@
...
+syntax (name := rintroPat.one) rcasesPat : rintroPat
+syntax (name := rintroPat.binder) "(" rintroPat+ (" : " term)? ")" : rintroPat
+syntax (name := rintro?) "rintro?" (" : " num)? : tactic
+syntax (name := rintro) "rintro" (ppSpace rintroPat)* (" : " term)? : tactic
+
+syntax (name := ext1) "ext1" (ppSpace rcasesPat)* : tactic
+syntax (name := ext1?) "ext1?" (ppSpace rcasesPat)* : tactic
+-- Implemented in Mathlib.Tactic.Ext
+-- syntax (name := ext) "ext" (ppSpace rcasesPat)* (" : " num)? : tactic
+syntax (name := ext?) "ext?" (ppSpace rcasesPat)* (" : " num)? : tactic
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Mathport/Syntax.lean:267]: Okay. The existing implementation of `ext` in mathlib4 doesn't use `rcasesPat`, and will need to be rewritten at some point. For now, I've just left comments at both places noting that there's a discrepancy.

I did move `rcasesPat` to a file `Mathlib.Tactic.Rcases`, which `Mathlib.Tactic.Ext` now imports (but doesn't yet use).

File: Mathlib/Mathport/Syntax.lean [hunk 6]
```diff
@@ -0,0 +1,737 @@
...
+syntax (name := inhabit) "inhabit " atomic(ident " : ")? term : tactic
+syntax (name := revertDeps) "revertDeps" (ppSpace (colGt ident))* : tactic
+syntax (name := revertAfter) "revertAfter " ident : tactic
+syntax (name := revertTargetDeps) "revertTargetDeps" : tactic
+syntax (name := clearValue) "clearValue" (ppSpace (colGt ident))* : tactic
+
+syntax (name := applyAssumption) "applyAssumption" : tactic
+-- Implemented in Mathlib.Tactic.SolveByElim
+-- syntax (name := solveByElim) "solveByElim" "*"? (" (" &"config" " := " term ")")?
+--   (&" only")? (" [" simpArg,* "]")? (" with " (colGt ident)+)? : tactic
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Mathport/Syntax.lean:337]: I'm having the same problem as with `normNum` below: when I declare the full `syntax`, then use `elab_rules` to implement just the initial fragment we have now, I get the `tactic ... has not been implemented` error.

File: Mathlib/Mathport/Syntax.lean (hunk 4 above)
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Mathport/Syntax.lean]: Solved.

File: Mathlib/Tactic/NormNum.lean [hunk 7]
```diff
@@ -138,14 +138,21 @@ def evalEq (α a b : Expr) : MetaM Expr := do
...
 end Meta
 
+namespace Tactic
+
+open Lean.Parser.Tactic in
+syntax (name := normNum) "normNum" (" [" simpArg,* "]")? (ppSpace location)? : tactic
+
 open Meta Elab.Tactic in
-elab "normNum" : tactic =>
+elab_rules : tactic | `(normNum) => do
```
Comment (digama0) [inline Mathlib/Tactic/NormNum.lean:147]:
```suggestion
elab_rules : tactic | `(tactic| normNum) => do
```
It is inferring the kind to `ident`, i.e. reading the syntax as a term `normNum` that doesn't exist and is maybe a local variable. @Kha Should the `: tactic` be used to influence the interpretation of syntax quotations in `inferMacroRulesAltKind`?

File: Mathlib/Tactic/LibrarySearch.lean [hunk 8]
```diff
@@ -92,8 +93,20 @@ def librarySearch (mvarId : MVarId) (lemmas : DiscrTree Name) (solveByElimDepth
...
+-- or requiring that a particular hypothesis is used in the solution, with `library_search using h`.
+syntax (name := librarySearch') "librarySearch" (" (" &"config" " := " term ")")?
+  (" [" simpArg,* "]")? (" with " (colGt ident)+)? (" using " (colGt ident)+)? : tactic
+
+-- For now we only implement the basic functionality.
+-- The full syntax is recognized, but will produce a "Tactic has not been implemented" error.
+
 open Elab.Tactic Elab Tactic in
-elab tk:"librarySearch" : tactic => do
+elab_rules : tactic | `(librarySearch%$tk) => do
```
Comment (digama0) [inline Mathlib/Tactic/LibrarySearch.lean:109]:
```suggestion
elab_rules : tactic | `(tactic| librarySearch%$tk) => do
```
This has the same issue. It was elaborating to `(librarySearch % $tk)`, i.e. the mod operator applied to two identifiers, so it definitely wouldn't have worked.

File: Mathlib/Tactic/NormNum.lean (hunk 7 above)
Comment (Kha) [reply Mathlib/Tactic/NormNum.lean:147]: It probably should, but it would require nontrivial changes in the parser.

**Comment by semorrison** [comment]
Great, thanks @digama0 for the explanation to use `tactic| ` in the pattern matches. I think this may be good to go now?

File: Mathlib/Mathport/Syntax.lean [hunk 9]
```diff
@@ -0,0 +1,727 @@
...
+end Tactic
+
+namespace Attr
+
+syntax (name := intro) "intro" : attr
+syntax (name := intro!) "intro!" : attr
+
+syntax (name := nolint) "nolint" (ppSpace ident)* : attr
+
+syntax (name := ext) "ext" (ppSpace ident)? : tactic
```
Comment (gebner) [inline Mathlib/Mathport/Syntax.lean:630]: This is not a tactic.  Also the ext attribute is never used with an argument in mathlib.
```suggestion
```

File: Mathlib/Mathport/Syntax.lean [hunk 10]
```diff
@@ -0,0 +1,727 @@
...
+syntax (name := squeezeDSimp?!) "squeezeDSimp?!" (" (" &"config" " := " term ")")? (&" only")?
+  (" [" simpArg,* "]")? (" with " (colGt ident)+)? (ppSpace location)? : tactic
+
+syntax (name := suggest) "suggest" (" (" &"config" " := " term ")")? (ppSpace num)?
+  (" [" simpArg,* "]")? (" with " (colGt ident)+)? (" using " (colGt ident)+)? : tactic
+-- Moved to Mathlib.Tactic.LibrarySearch
+-- syntax (name := librarySearch) "librarySearch" (" (" &"config" " := " term ")")?
+--   (" [" simpArg,* "]")? (" with " (colGt ident)+)? (" using " (colGt ident)+)? : tactic
+syntax (name := librarySearch!) "librarySearch!" (" (" &"config" " := " term ")")?
+  (" [" simpArg,* "]")? (" with " (colGt ident)+)? (" using " (colGt ident)+)? : tactic
```
Comment (gebner) [inline Mathlib/Mathport/Syntax.lean:456]:
```suggestion
```

These transient tactics are not (supposed to be!) used in mathlib and they will almost certainly look different in Lean 4, I really don't think we should spend time on porting their syntax 1-to-1.

We could also append `3` to all names, then the syntax doesn't matter and we can just remove the dangling syntax after the port.

But this is not a blocker, none of the syntaxes clash now.

File: Mathlib/Tactic/LibrarySearch.lean [hunk 11]
```diff
@@ -92,8 +92,20 @@ def librarySearch (mvarId : MVarId) (lemmas : DiscrTree Name) (solveByElimDepth
...
+-- or requiring that a particular hypothesis is used in the solution, with `library_search using h`.
+syntax (name := librarySearch') "librarySearch" (" (" &"config" " := " term ")")?
+  (" [" simpArg,* "]")? (" with " (colGt ident)+)? (" using " (colGt ident)+)? : tactic
+
+-- For now we only implement the basic functionality.
+-- The full syntax is recognized, but will produce a "Tactic has not been implemented" error.
+
 open Elab.Tactic Elab Tactic in
-elab tk:"librarySearch" : tactic => do
+elab_rules : tactic | `(tactic| librarySearch%$tk) => do
```
Comment (gebner) [inline Mathlib/Tactic/LibrarySearch.lean:108]: Can we just translate the Lean 3 `library_search` to `librarySearch3` instead so that we don't have to match the syntax?

File: Mathlib/Mathport/Syntax.lean [hunk 12]
```diff
@@ -0,0 +1,727 @@
...
+syntax (name := unfold) "unfold" (" (" &"config" " := " term ")")?
+  (ppSpace (colGt ident))* (ppSpace location)? : tactic
+syntax (name := unfold1) "unfold1" (" (" &"config" " := " term ")")?
+  (ppSpace (colGt ident))* (ppSpace location)? : tactic
+syntax (name := inferOptParam) "inferOptParam" : tactic
+syntax (name := inferAutoParam) "inferAutoParam" : tactic
+syntax (name := guardExprEq) "guardExpr " term:51 " =ₐ " term : tactic
+syntax (name := guardTarget) "guardTarget" " =ₐ " term : tactic
+syntax (name := guardHyp) "guardHyp " ident
+  ((" : " <|> " :ₐ ") term)? ((" := " <|> " :=ₐ ") term)? : tactic
```
Comment (gebner) [inline Mathlib/Mathport/Syntax.lean:214]: Also duplicates.

File: Mathlib/Mathport/Syntax.lean [hunk 13]
```diff
@@ -0,0 +1,727 @@
...
+syntax (name := matchTarget) "matchTarget " term : tactic
+syntax (name := byCases) "byCases " atomic(ident " : ")? term : tactic
```
Comment (gebner) [inline Mathlib/Mathport/Syntax.lean:216]: Duplicate.

File: Mathlib/Mathport/Syntax.lean [hunk 14]
```diff
@@ -0,0 +1,727 @@
...
+  (ppSpace (colGt ident))* (ppSpace location)? : tactic
+syntax (name := unfold1) "unfold1" (" (" &"config" " := " term ")")?
+  (ppSpace (colGt ident))* (ppSpace location)? : tactic
+syntax (name := inferOptParam) "inferOptParam" : tactic
+syntax (name := inferAutoParam) "inferAutoParam" : tactic
+syntax (name := guardExprEq) "guardExpr " term:51 " =ₐ " term : tactic
+syntax (name := guardTarget) "guardTarget" " =ₐ " term : tactic
+syntax (name := guardHyp) "guardHyp " ident
+  ((" : " <|> " :ₐ ") term)? ((" := " <|> " :=ₐ ") term)? : tactic
+syntax (name := matchTarget) "matchTarget " term : tactic
```
Comment (gebner) [inline Mathlib/Mathport/Syntax.lean:215]: Already in Tactic.Basic.

**Review by gebner — COMMENTED** [review]
There's still quite a few duplicate clashing tactics left.

File: Mathlib/Mathport/Syntax.lean (hunk 10 above)
Comment (digama0) [reply Mathlib/Mathport/Syntax.lean:456]: I would prefer that the syntaxes continue to exist, although as you say they can be renamed if you want them to look different. But why wouldn't we want them to have a grammar like this? If someone is using `librarySearch!` they shuold be able to add all these config options and so on, I don't see a compelling reason not to have the tactic and a roughly equivalent syntax, not for porting but for lean 3 parity.

File: Mathlib/Mathport/Syntax.lean (hunk 9 above)
Comment (digama0) [reply Mathlib/Mathport/Syntax.lean:630]:
```suggestion
syntax (name := ext) "ext" (ppSpace ident)? : attr
```
I know it's not used with a name right now, but this is part of the original tactic. Note that it is already simplified from the original syntax which used a list of names and `(->)`, this was PR leanprover-community/mathlib#8785 . But the name argument is actually useful in principle and I could not find a reason to drop the feature, although porting will be trivial since the feature isn't used.

In other words, this is a mathlib4 todo list for us to design the lean 4 tactic language we want to support with attention to lean 3 parity, not just a basis for synport. Removing the argument and erroring on any use of the with-argument form in synport is an option, but I don't want such decreases in functionality to sneak up on us without at least a quorum of some kind (ideally including the original author of the tactic, who likely put the feature in for a reason).

File: Mathlib/Tactic/LibrarySearch.lean (hunk 11 above)
Comment (digama0) [reply Mathlib/Tactic/LibrarySearch.lean:108]: Is there a reason that `by librarySearch [args]` shouldn't work in lean 4? I understand that it's not done yet but I want all the lean 3 tactics to have equivalents in lean 4.

File: Mathlib/Mathport/Syntax.lean (hunk 12 above)
Comment (digama0) [reply Mathlib/Mathport/Syntax.lean:214]: The syntax is a little different. We should figure out what syntax we want to use here. The original syntax evolved over time as more variations were added; I made up a syntax that generalizes well to the variations here, and we would need to update mathlib4 to match if this looks good.

Comment (digama0) [reply Mathlib/Mathport/Syntax.lean:214]: In fact, the mathlb4 tactic seems to be asking itself the same question:
```lean
--TODO : which expr equality to use?
elab "guardExprEq " r:term " := " p:term : tactic => withMainContext do
```
The proposed syntax answers the question thus:
```lean
syntax (name := guardExprEq) "guardExpr " term:51 " =ₐ " term : tactic -- alpha equal
syntax (name := guardExprStrict) "guardExpr " term:51 " == " term : tactic -- syntax equal
syntax (name := guardExprEq') "guardExpr " term:51 " = " term : tactic -- defeq
```

File: Mathlib/Mathport/Syntax.lean (hunk 13 above)
Comment (digama0) [reply Mathlib/Mathport/Syntax.lean:216]: Not a match, and that one's in core so it is more problematic to override. Probably this one needs to be renamed to `byCases'`

File: Mathlib/Tactic/LibrarySearch.lean (hunk 11 above)
Comment (gebner) [reply Mathlib/Tactic/LibrarySearch.lean:108]: No that should still work, as well as the `using` argument that is also on the todo list.  The configuration object is almost certainly going to change (the lean 3 one refers to the built-in apply tactic).  We might also want to replace solve-by-elim with Jannis' auto port as discussed on Monday, affecting the `with` syntax.

Most importantly though, aligning these tactics is completely unnecessary.

File: Mathlib/Mathport/Syntax.lean (hunk 10 above)
Comment (gebner) [reply Mathlib/Mathport/Syntax.lean:456]: As discussed in the other comment, the functionality should certainly be available in Lean 4 but there's no point in spending any effort on porting them since they don't occur in mathlib.

Since you've mentioned the bang-version of librarySearch.  The bang-less version (doing only reducible matching) is an optimization we had to do in Lean 3 because semireducible matching was too slow.  We can hopefully eliminate this variant entirely in Lean 4 using better indexing.

File: Mathlib/Mathport/Syntax.lean (hunk 9 above)
Comment (gebner) [reply Mathlib/Mathport/Syntax.lean:630]: The reason for the extra argument is afaict that the Lean 3 version of `ext` indexes the lemmas by the constant name of the head symbol.  The Lean 4 version uses discrimination trees for indexing and doesn't support this annotation.  (And it shouldn't, the lemma indexing and lookup done by the old ext tactic is a huge hack.)

File: Mathlib/Mathport/Syntax.lean (hunk 14 above)
Comment (semorrison) [reply Mathlib/Mathport/Syntax.lean:215]: @digama0, could you explain your thinking about the `=ₐ` here? It seems contrary to the principle of preserving mathlib3 syntax.

File: Mathlib/Mathport/Syntax.lean (hunk 12 above)
Comment (semorrison) [reply Mathlib/Mathport/Syntax.lean:214]: Fixed `guardExpr`, `guardTarget`, and `guardHyp`, and added comments about things that still need to be implemented.

File: Mathlib/Mathport/Syntax.lean (hunk 14 above)
Comment (semorrison) [reply Mathlib/Mathport/Syntax.lean:215]: Done.

File: Mathlib/Mathport/Syntax.lean (hunk 13 above)
Comment (semorrison) [reply Mathlib/Mathport/Syntax.lean:216]: Renamed, and added a comment explaining the difference.

**Comment by semorrison** [comment]
Could we just punt for now on the transient tactics? I don't think having the syntax statements here for them causes any problem, and we can delete them / modify them / whatever in later PRs. I'd prefer to get this PR out of the way so we can finish off providing CI and a downloadable artifact for mathport!

**Comment by gebner** [comment]
> Could we just punt for now on the transient tactics? I don't think having the syntax statements here for them causes any problem, and we can delete them / modify them / whatever in later PRs.

I completely agree with you, that's why I said earlier:

> But this is not a blocker, none of the syntaxes clash now.

bors r+
